# 问题1
## 5.1 alignment
1. 目的：将质控后的reads比对到目的基因组上
2. 使用软件： BWA-MEM or Bowtie2，本流程采用`Bowtie2`  

3. 代码：  

```bash
mkdir -p /mnt/d/ATAC/alignment
bowtie2_index=/mnt/d/ATAC/genome/mm10
align_dir=/mnt/d/ATAC/alignment

# 循环 
cd /mnt/d/ATAC/trim2/ 
cat >config.raw <<EOF
SRR11539111.fq.gz  SRR11539111_1_val_1.fq.gz  SRR11539111_2_val_2.fq.gz
SRR11539112.fq.gz  SRR11539112_1_val_1.fq.gz  SRR11539112_2_val_2.fq.gz
SRR11539115.fq.gz  SRR11539115_1_val_1.fq.gz  SRR11539115_2_val_2.fq.gz
SRR11539116.fq.gz  SRR11539116_1_val_1.fq.gz  SRR11539116_2_val_2.fq.gz
EOF
```
### 最开始没有加-X 2000参数(与核小体解旋测得的个数有关=400bp✖n)，导致paired比对率很低
```bash
cat config.raw | while read id;
do echo $id 
  arr=($id)
  fq1=${arr[1]}
  fq2=${arr[2]}
  sample=${arr[0]}
  bowtie2  -p 6  -x  $bowtie2_index  -1  $fq1 -2 $fq2 \
  2>$align_dir/${sample}.summary \
  -S $align_dir/${sample}.sam
done
```
* 结果：
```bash
/mnt/d/ATAC/align$ samtools flagstat SRR11539111.sort.bam
98013300 + 0 in total (QC-passed reads + QC-failed reads)
98013300 + 0 primary
0 + 0 secondary
0 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
95930542 + 0 mapped (97.88% : N/A)
95930542 + 0 primary mapped (97.88% : N/A)
98013300 + 0 paired in sequencing
49006650 + 0 read1
49006650 + 0 read2
84125536 + 0 properly paired (85.83% : N/A) #看这个数据
94651430 + 0 with itself and mate mapped
1279112 + 0 singletons (1.31% : N/A)
785374 + 0 with mate mapped to a different chr
288053 + 0 with mate mapped to a different chr (mapQ>=5)
```
### 加上-X 2000参数
```bash
cat config.raw | while read id;
do echo $id 
  arr=($id)
  fq1=${arr[1]}
  fq2=${arr[2]}
  sample=${arr[0]}
  bowtie2  -p 6  -x  $bowtie2_index --very-sensitive -X 2000 -1  $fq1 -2 $fq2 \
  2>$align_dir/${sample}.summary \
  -S $align_dir/${sample}.sam
done
```
* 结果：

```bash
# 以SRR11539111为例
# summmary
49006650 reads; of these:
  49006650 (100.00%) were paired; of these:
    1643074 (3.35%) aligned concordantly 0 times
    33537190 (68.43%) aligned concordantly exactly 1 time
    13826386 (28.21%) aligned concordantly >1 times
    ----
    1643074 pairs aligned concordantly 0 times; of these:
      306783 (18.67%) aligned discordantly 1 time
    ----
    1336291 pairs aligned 0 times concordantly or discordantly; of these:
      2672582 mates make up the pairs; of these:
        1573243 (58.87%) aligned 0 times
        664621 (24.87%) aligned exactly 1 time
        434718 (16.27%) aligned >1 times
98.39% overall alignment rate

# samtools flagstat
/mnt/d/ATAC/alignment$ samtools flagstat SRR11539111.fq.gz.sam
98013300 + 0 in total (QC-passed reads + QC-failed reads)
98013300 + 0 primary
0 + 0 secondary
0 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
96440057 + 0 mapped (98.39% : N/A)
96440057 + 0 primary mapped (98.39% : N/A)
98013300 + 0 paired in sequencing
49006650 + 0 read1
49006650 + 0 read2
94727152 + 0 properly paired (96.65% : N/A)
95584080 + 0 with itself and mate mapped
855977 + 0 singletons (0.87% : N/A)
160994 + 0 with mate mapped to a different chr
89323 + 0 with mate mapped to a different chr (mapQ>=5)
```
可以看到，添加了-X命令之后properly paired比例大大上升。

# 问题2:
## 6.1 remove_PCR-duplicate_reads
目的：去除因为PCR偏好性导致的reads重复扩增  

```bash
mkdir -p /mnt/d/ATAC/rmdup
cd /mnt/d/ATAC/alignment

cat config.raw | while read id;
do echo $id 
  arr=($id)
  sample=${arr[0]}

  java -Dpicard.useLegacyParser=false \
  -jar /mnt/d/biosoft/picard/picard.jar \
     MarkDuplicates -I ${sample}.sort.bam \
	  -O ../rmdup/${sample}.rmdup.bam \
	 -REMOVE_DUPLICATES true \
   -VALIDATION_STRINGENCY LENIENT \
	 -M ../rmdup/${sample}.log  
   
   samtools index -@ 7 ../rmdup/${sample}.rmdup.bam
   samtools flagstat -@ 7 ../rmdup/${sample}.rmdup.bam > ../rmdup/${sample}.rmdup.stat
done
```
### 最开始true拼错了！！！
### 所以只标记了dup并没有去除，原结果存放在/mnt/d/ATAC/rmdup/nonrmdup

# 问题3  
# 8.call_peaks
## 最开始peaks/代码

```bash
macs2 callpeak  -g mm -f BEDPE --nomodel  \
   --shift -100 --extsize 200 -n ${sample} -t ./${sample}.Tn5.bedpe \
  --outdir ../peaks1/
```
!当有-f BEDPE，paired-end模式开启后，会自动舍弃--shift 和 --extsize参数，所以peaks1和peaks2结果几乎一样。  


## 与/peaks相比，/peaks1加了几个参数，删除了shift和extsize
cp /mnt/d/ATAC/rmdup/config.raw /mnt/d/ATAC/shifted/config.raw
mkdir -p /mnt/d/ATAC/peaks1/
cat config.raw | while read id;
do echo $id 
  arr=($id)
  sample=${arr[0]}

  macs2 callpeak  -g mm -f BEDPE --nomodel --keep-dup all \
   --cutoff-analysis -n ${sample} -t ./ ${sample}.Tn5.bedpe \
  --outdir ../peaks1/ 
done
## 之后的流程都使用peaks1数据
