# 问题1
## 5.1 alignment
1. 目的：将质控后的reads比对到目的基因组上
2. 使用软件： BWA-MEM or Bowtie2，本流程采用`Bowtie2`  

3. 代码：  

```bash
mkdir -p /mnt/d/ATAC/alignment
bowtie2_index=/mnt/d/ATAC/genome/mm10
align_dir=/mnt/d/ATAC/alignment

# 循环 
cd /mnt/d/ATAC/trim2/ 
cat >config.raw <<EOF
SRR11539111.fq.gz  SRR11539111_1_val_1.fq.gz  SRR11539111_2_val_2.fq.gz
SRR11539112.fq.gz  SRR11539112_1_val_1.fq.gz  SRR11539112_2_val_2.fq.gz
SRR11539115.fq.gz  SRR11539115_1_val_1.fq.gz  SRR11539115_2_val_2.fq.gz
SRR11539116.fq.gz  SRR11539116_1_val_1.fq.gz  SRR11539116_2_val_2.fq.gz
EOF
```
### 最开始没有加-X 2000参数(与核小体解旋测得的个数有关=400bp✖n)，导致paired比对率很低
```bash
cat config.raw | while read id;
do echo $id 
  arr=($id)
  fq1=${arr[1]}
  fq2=${arr[2]}
  sample=${arr[0]}
  bowtie2  -p 6  -x  $bowtie2_index  -1  $fq1 -2 $fq2 \
  2>$align_dir/${sample}.summary \
  -S $align_dir/${sample}.sam
done
```
* 结果：
```bash
/mnt/d/ATAC/align$ samtools flagstat SRR11539111.sort.bam
98013300 + 0 in total (QC-passed reads + QC-failed reads)
98013300 + 0 primary
0 + 0 secondary
0 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
95930542 + 0 mapped (97.88% : N/A)
95930542 + 0 primary mapped (97.88% : N/A)
98013300 + 0 paired in sequencing
49006650 + 0 read1
49006650 + 0 read2
84125536 + 0 properly paired (85.83% : N/A) #看这个数据
94651430 + 0 with itself and mate mapped
1279112 + 0 singletons (1.31% : N/A)
785374 + 0 with mate mapped to a different chr
288053 + 0 with mate mapped to a different chr (mapQ>=5)
```
### 加上-X 2000参数
```bash
cat config.raw | while read id;
do echo $id 
  arr=($id)
  fq1=${arr[1]}
  fq2=${arr[2]}
  sample=${arr[0]}
  bowtie2  -p 6  -x  $bowtie2_index --very-sensitive -X 2000 -1  $fq1 -2 $fq2 \
  2>$align_dir/${sample}.summary \
  -S $align_dir/${sample}.sam
done
```
* 结果：

```bash
# 以SRR11539111为例
# summmary
49006650 reads; of these:
  49006650 (100.00%) were paired; of these:
    1643074 (3.35%) aligned concordantly 0 times
    33537190 (68.43%) aligned concordantly exactly 1 time
    13826386 (28.21%) aligned concordantly >1 times
    ----
    1643074 pairs aligned concordantly 0 times; of these:
      306783 (18.67%) aligned discordantly 1 time
    ----
    1336291 pairs aligned 0 times concordantly or discordantly; of these:
      2672582 mates make up the pairs; of these:
        1573243 (58.87%) aligned 0 times
        664621 (24.87%) aligned exactly 1 time
        434718 (16.27%) aligned >1 times
98.39% overall alignment rate

# samtools flagstat
/mnt/d/ATAC/alignment$ samtools flagstat SRR11539111.fq.gz.sam
98013300 + 0 in total (QC-passed reads + QC-failed reads)
98013300 + 0 primary
0 + 0 secondary
0 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
96440057 + 0 mapped (98.39% : N/A)
96440057 + 0 primary mapped (98.39% : N/A)
98013300 + 0 paired in sequencing
49006650 + 0 read1
49006650 + 0 read2
94727152 + 0 properly paired (96.65% : N/A)
95584080 + 0 with itself and mate mapped
855977 + 0 singletons (0.87% : N/A)
160994 + 0 with mate mapped to a different chr
89323 + 0 with mate mapped to a different chr (mapQ>=5)
```
可以看到，添加了-X命令之后properly paired比例大大上升。

# 问题2:
## 6.1 remove_PCR-duplicate_reads
目的：去除因为PCR偏好性导致的reads重复扩增  

```bash
mkdir -p /mnt/d/ATAC/rmdup
cd /mnt/d/ATAC/alignment

cat config.raw | while read id;
do echo $id 
  arr=($id)
  sample=${arr[0]}

  java -Dpicard.useLegacyParser=false \
  -jar /mnt/d/biosoft/picard/picard.jar \
     MarkDuplicates -I ${sample}.sort.bam \
	  -O ../rmdup/${sample}.rmdup.bam \
	 -REMOVE_DUPLICATES true \
   -VALIDATION_STRINGENCY LENIENT \
	 -M ../rmdup/${sample}.log  
   
   samtools index -@ 7 ../rmdup/${sample}.rmdup.bam
   samtools flagstat -@ 7 ../rmdup/${sample}.rmdup.bam > ../rmdup/${sample}.rmdup.stat
done
```
### 最开始true拼错了！！！
### 所以只标记了dup并没有去除，原结果存放在/mnt/d/ATAC/rmdup/nonrmdup

# 问题3  
# 8.call_peaks
## 最开始peaks/代码

```bash
macs2 callpeak  -g mm -f BEDPE --nomodel  \
   --shift -100 --extsize 200 -n ${sample} -t ./${sample}.Tn5.bedpe \
  --outdir ../peaks1/
```
!当有-f BEDPE，paired-end模式开启后，会自动舍弃--shift 和 --extsize参数，所以peaks1和peaks2结果几乎一样。  


## 与/peaks相比，/peaks1加了几个参数，删除了shift和extsize
cp /mnt/d/ATAC/rmdup/config.raw /mnt/d/ATAC/shifted/config.raw
mkdir -p /mnt/d/ATAC/peaks1/
cat config.raw | while read id;
do echo $id 
  arr=($id)
  sample=${arr[0]}

  macs2 callpeak  -g mm -f BEDPE --nomodel --keep-dup all \
   --cutoff-analysis -n ${sample} -t ./ ${sample}.Tn5.bedpe \
  --outdir ../peaks1/ 
done
## 之后的流程都使用peaks1数据



# 问题3
## 最开始先call peak，再去除的black list region，记录一下前期没有去除黑色名单call的peaks数
#### 8. call_peaks 
1. 目的： 下一步需要在统计学上判断真实的peak，因为Tn5在染色体上结合是个概率事件，如何判断这个位置的reads足够为一个peak，这就需要用到统计检测。ATAC-seq 数据分析的第二个主要步骤是识别开放区域（也称为 Peak），后续高级分析以此为基础。  

2. 软件：目前，`MACS2` 是 ENCODE ATAC-seq 流程的默认 Peak caller 程序。  

3. !!!重要：关于是否使用[-f BEDPE的讨论](https://github.com/macs3-project/MACS/issues/331)，可根据需要选择合适的callpeak参数。  


4. 其他： 


* ATAC-seq关心的是在哪里切断，断点才是peak的中心，所以使用shift模型，--shift -75或-100.   

* 这里选用固定宽度（fixed-width）的peaks,优点有：   
1）对大量的peaks进行counts和motif分析时可以减小误差；  
2）对于大量数据集的可以合并峰得到一致性的peaks;   

* 一个样本的overlaps他们是通过迭代移除的方法，首先保留最显著的peak，然后任何与最显著peak有直接overlap的peaks都被移除；接着对另一个最显著性的peak进行相同的操作，最终保留所有更显著的peaks，移除与其有直接overlaps的peaks  
* 注：后续分析过程需要用到IDR提取consensus peak，建议MACS2 callpeaks的步骤参数设置不要过于严格，以便鉴定出更多的peaks。

4. 代码：
```bash
mkdir -p /mnt/d/ATAC/peaks/
cd /mnt/d/ATAC/shifted/

# 注：本流程使用的是经过转化的bedpe
# 单个样本
macs2 callpeak  -g mm -f BEDPE --nomodel \
  -n SRR11539111 -t ./SRR11539111.Tn5.bedpe \
  --outdir /mnt/d/ATAC/peaks

# 循环
ls *.Tn5.bedpe| while read id; do
  macs2 callpeak  -g mm -f BEDPE --nomodel \
   -n $id -t $id \
  --outdir ../peaks/ 
done

# 推荐使用peaks1代码  
# 与/peaks相比，/peaks1加了几个参数，删除了shift和extsize
cp /mnt/d/ATAC/rmdup/config.raw /mnt/d/ATAC/shifted/config.raw
mkdir -p /mnt/d/ATAC/peaks1/
cat config.raw | while read id;
do echo $id 
  arr=($id)
  sample=${arr[0]}

  macs2 callpeak  -g mm -f BEDPE --nomodel --keep-dup all \
   --cutoff-analysis -n ${sample} -t ./${sample}.Tn5.bedpe \
  --outdir ../peaks1/ 
done

# 如果用的不是专门双端测序的bedpe，而是bed文件，采用下面代码
# 单个样本
cd /mnt/d/ATAC/shifted/
macs2 callpeak  -g mm --nomodel \
  --shift -100 --extsize 200 -n SRR11539111 -t ./SRR11539111.bed \
  --outdir /mnt/d/ATAC/peaks

# 循环
mkdir -p /mnt/d/ATAC/peaks2
cd /mnt/d/ATAC/shifted/
ls *.bed| while read id; do
  macs2 callpeak  -g mm --nomodel \
  --shift -100 --extsize 200 -n $id -t $id \
  --outdir ../peaks2/ 
done

或者

cp /mnt/d/ATAC/rmdup/config.raw /mnt/d/ATAC/shifted/config.raw
cat config.raw | while read id;
do echo $id 
  arr=($id)
  sample=${arr[0]}

  macs2 callpeak  -g mm --nomodel \
  --shift -100 --extsize 200 -n ${sample} -t ./${sample}.Tn5.bed \
  --outdir ../peaks2/ 
done
```
* macs2 callpeaks [参数](https://manpages.ubuntu.com/manpages/impish/man1/macs2_callpeak.1.html#:~:text=Please%20note%20that%20if%20the%20format%20is%20set,predicting%20the%20fragment%20size%20first%20and%20extending%20reads.)   

```bash
-t bed文件;  -g 比对基因组;  -n 前缀;

-Q/–QVALUE：qvalue (minimum FDR)设定call significant regions的阈值；默认，0.01； 

--nomodel的意思是让其不要建立双峰模型来使两个“相邻”的峰shift成一个峰，而是向外shift(也就是在nomodel后要加上--shift -75 --extsize 150的参数);  

-P/–PVALUE：设定p值时， qvalue不再起作用。  

–NOMODEL：MACS 不构建模型。  

–EXTSIZE：设定–nomodel，MACS 会沿着 5’->3’方向延伸reads；如果转录因子结合区域长200bp，
把所有的reads都统一成200bp长，并且移动了100bp，这样就保证了这200bp的中心是酶切位点.  

–SHIFT：–shiftsize已经被 –extsize所替代；–nomodel设定之后，MACS 会用这个参数剪切reads5’，利用–extsize 延伸reads 3’端；
如果设为负数，方向相反(3’->5’ );
ChIP-Seq建议设置为0；当检测富集切割位点时，例如DNAseI-Seq datasets，
此参数应该设为 -1 * half of EXTSIZE( EXTSIZE设为200，此参数为-100).  

-f      {AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE,BEDPE},       --format
       {AUTO,BAM,SAM,BED,ELAND,ELANDMULTI,ELANDEXPORT,BOWTIE,BAMPE,BEDPE}
              Format  of  tag  file, "AUTO", "BED" or "ELAND" or "ELANDMULTI" or "ELANDEXPORT" or
              "SAM" or "BAM" or "BOWTIE" or "BAMPE" or "BEDPE". The default AUTO option will  let
              MACS  decide  which  format  (except for BAMPE and BEDPE which should be implicitly
              set) the file is. Please check the definition in README. Please note  that  if  the
              format  is  set  as  BAMPE or BEDPE, MACS2 will call its special Paired-end mode to
              call peaks by piling up the actual ChIPed fragments defined by both  aligned  ends,
              instead of predicting the fragment size first and extending reads. Also please note
              that the BEDPE only contains three columns, and is NOT the same BEDPE  format  used
              by BEDTOOLS.  DEFAULT: "AUTO"    
--nolambda: 如果为真，MACS 将使用固定背景 lambda 作为每个峰区域的本地 lambda。 
             通常，MACS 计算动态局部 lambda 以反映由于潜在的染色质可及性而导致的局部偏差。  
--cutoff-analysis
              While set, MACS2 will analyze number or total length of peaks that can be called by
              different p-value cutoff then output a summary table to help user decide  a  better
              cutoff.  The table will be saved in NAME_cutoff_analysis.txt file. Note, minlen and
              maxgap may affect the results. WARNING: May take ~30 folds longer time  to  finish.
              The  result  can  be useful for users to decide a reasonable cutoff value. DEFAULT:
              False
```

* 两个例子：  

DNAse-Seq，想将平滑窗口设为200bps时，使用参数‘–nomodel –shift -100 –extsize 200’。  
nucleosome-seq，使用核小体一半大小进行小波分析获得核小体中心的峰；当缠绕核小体DNA长度为147bps，可使用参数‘–nomodel –shift 37 –extsize 73’。  

链接：https://www.imooc.com/article/270403  





* 结果： 
最终生成三个文件：narrowpeak, peaks.xls,summits.bed，[详细解释](https://github.com/hbctraining/In-depth-NGS-Data-Analysis-Course/blob/master/sessionV/lessons/04_peak_calling_macs.md)  

`_peaks.narrowPea`k：BED6+4格式文件，其中包含峰值位置以及峰值峰值，p值和q值  
`_peaks.xls`：包含有关调用峰的信息的表格文件。其他信息包括堆积和折叠富集  
`_summits.bed`：每个山峰的峰顶位置。要找到结合位点的基序，建议使用此文件  
1. narrowpeak  
```bash
chr1	3670812	3672021	SRR11539111_peak_1	125	.	5.29967	15.80579	12.56397	799
chr1	4785447	4785897	SRR11539111_peak_2	153	.	6.79790	18.76872	15.35057	249
# 染色体 起始位点 结束位点（Tn5转化过）peak名称  score   链 signal_value（fold-change）  -log10(pvalue) -log10qvalue 峰位与peak起点的距离
```
2. peaks.xls
```bash
# chr	start   	end 	length	abs_summit	pileup位置堆积信号  	-log10(pvalue)  fold_enrichment	-log10(qvalue)	name
chr1	3670813	3672021	1209	3671612	33	15.80579	5.29967	12.56397	SRR11539111_peak_1
```
3. summits.bed
```bash
chr1    3671611 3671612 SRR11539111_peak_1      12.56397
chr1    4785696 4785697 SRR11539111_peak_2      15.35057
#-log10pvalue
```
* bed_bedpe结果比较
```bash
# bedpe
wc -l SRR11539111_peaks.narrowPeak
17001

# bed
28552
# bedpe的长度会更长一点，后续分析都采用bedpe
```

